src/objects/CellContainer.cpp

```diff
@@ -0,0 +1,346 @@
+#include "CellContainer.h"
+#include "utils/Arguments.h"
+#include "utils/ArrayUtils.h"
+#include "utils/CLIUtils.h"
+#include "utils/StringUtils.h"
+#include <cassert>
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <spdlog/spdlog.h>
+#include <vector>
+
+/* constructor */
+CellContainer::CellContainer(const std::array<double, 3> &domainSize,
+                             const std::array<BoundaryCondition, 6> &conditions, double cutoff,
+                             ParticleContainer &particles, size_t dim)
+    : domainSize{domainSize}, conditions{conditions}, cutoff{cutoff}, particles{particles} {
+    // check correct dimensions (could probably be a boolean instead...)
+    if (dim < 2 || dim > 3)
+        CLIUtils::error("Invalid cell container dimensions! (must be 2 or 3)", StringUtils::fromNumber(dim));
+
+    // check that domain size and cutoff are initialized
+    if (cutoff == INF)
+        CLIUtils::error("Cutoff radius not initialized!");
+    if (domainSize[0] == INF || domainSize[1] == INF || domainSize[2] == INF)
+        CLIUtils::error("Domain size not initialized!");
+
+    SPDLOG_TRACE("Generating CellContainer with domain size {} and cutoff radius {} (in {} dimensions)",
+                 ArrayUtils::to_string(domainSize), cutoff, dim);
+
+    // determining cell size
+    for (size_t i = 0; i < dim; i++) {
+        if (std::fabs(std::fmod(domainSize[i], cutoff)) < 1e-9) {
+            // perfect fit
+            cellSize[i] = cutoff;
+            numCells[i] = std::round(domainSize[i] / cellSize[i]) + 2;
+            SPDLOG_DEBUG("Cutoff perfectly divides domain size, using cutoff radius {} as cell size for dimension {}.",
+                         cutoff, i);
+        } else {
+            // we round up cellsize so that cell > cutoff and we only search adjacent cells
+            cellSize[i] = domainSize[i] / std::floor(domainSize[i] / cutoff);
+            numCells[i] = std::round(domainSize[i] / cellSize[i]) + 2;

```

Comment:
feels more readable and avoids rounding twice (once with floor and once with round)
```suggestion
            numCells[i] = std::floor(domainSize[i] / cellSize[i]);
            cellSize[i] = domainSize[i] / numCells[i] ;
            numCells[i] += 2; // 2 halo cells
```